cohortextractor/emis_backend.py

```diff
@@ -1025,6 +1030,52 @@ def validate_recent_date(self, date, max_delta_days=30):
             msg = f"{self._current_column_name} must be passed a date more recent than {max_delta_days} days in the past"
             raise ValueError(msg)
 
+    def get_aggregate_expression(
+        self, column_type, column_definitions, column_names, aggregate_function
+    ):
+        """Return an expression that is used to find the maximum or minimum value across
+        a number of other given columns.
+
+        We cannot use Table Value Constructors as in the TPP backend (this gives a
+        cryptic error message: "Given correlated subquery is not supported").
+
+        Instead, we use GREATEST() or LEAST(), but we have to be careful to handle
+        correctly any arguments to GREATEST/LEAST that have replaced NULLs.  To do this,
+        we replace any occurences of the default value for the given column_type with
+        the most extreme value possible for the column_type.  (So when finding the
+        maximum, we replace the default value with a small value, and when finding the
+        minimum, with a large value.)
+
+        If all the arguments to GREATEST/LEAST have replaced NULLs, GREATEST/LEAST will
+        return the extreme value, so when that happens, we have to replace this with the
+        default value for the column type.
+        """
+
+        default_value = quote(self.get_default_value_for_type(column_type))
+        function = {"MAX": "GREATEST", "MIN": "LEAST"}[aggregate_function]
+
+        if column_type in ["int", "float"]:
+            extreme_value_lookup = {"MAX": -(2 ** 63), "MIN": 2 ** (63 - 1)}
+            extreme_value = [aggregate_function]
+        elif column_type == "date":
+            extreme_value_lookup = {"MAX": "'0001-01-01'", "MIN": "'9999-12-31'"}
+        else:
+            assert False, column_type
+        extreme_value = extreme_value_lookup[aggregate_function]
+
+        components = ", ".join(
+            f"""
+            CASE WHEN {column_definitions[name]} = {default_value}
+                THEN {extreme_value}
+            ELSE {column_definitions[name]} END"""
+            for name in column_names
+        )
+
+        return f"""
+        CASE WHEN {function}({components}) = {extreme_value}
+            THEN {default_value}
+        ELSE {function}({components}) END"""
+
```

Comment:
So if I've understood this right, the single edge case that would be a bug is where there is one or more genuine values of `2 ** 63` that we want to find the `MIN` of, and vice-versa? 

Given this is exceedingly unlikely it seems the cost of adding a check exceeds the benefits, but I do wonder if a note should be added to the documentation somewhere, or a warning in the docstring?

Reply:
I don't think you have understood this right.  However, writing this all out has revealed a bug.

Firstly: we don't expect to handle having genuine values of `2 ** 63` at all.

Suppose we have the following pairs of values for three patients, and we want to find the minimum of each pair.

| id | value1 | value2 |
| - | - | - |
| 1 | 100 | 200 |
| 2 | 100 | null |
| 3 | null | null |
| 4 | 100 | 0 |

Here are the expected results:

| id | value1 | value2 | exp_min |
| - | - | - | - |
| 1 | 100 | 200 | 100 |
| 2 | 100 | null | 100 |
| 3 | null | null | 0 |
| 4 | 100 | 0 | 0 |

(We want a zero for row 3, because we replace nulls with a default value, and the default value for an integer column is zero.)

Earlier on in the process, we have already converted nulls to the default value, so when the SQL for this function is executed, we have:

| id | value1 | value2 | value1' | value2' | min(value1', value2') | exp_min
| - | - | - | - | - | - | -
| 1 | 100 | 200 | 100 | 200 | 100 | 100
| 2 | 100 | null | 100 | 0 | 0 | 100
| 3 | null | null | 0 | 0 | 0 | 0
| 4 | 100 | 0 | 100 | 0 | 0 | 0

Finding `min(value1', value2')` will give the wrong result for row 2.

So instead, we convert any default values to a large number:
| id | value1 | value2 | value1' | value2' | value1'' | value2'' | min(value1'', value2'') | exp_min
| - | - | - | - | - | - | - | - | -
| 1 | 100 | 200 | 100 | 200 | 100 | 200 | 100 | 100
| 2 | 100 | null | 100 | 0 | 100 | 9999 | 100 | 100
| 3 | null | null | 0 | 0 | 9999 | 9999 | 9999 | 0
| 4 | 100 | 0 | 100 | 0 | 100 | 9999 | 100 | 0

`min(value1'', value2'')` now gives the right result for row 2, but not for row 3 (which we fix) or row 4 (which we don't fix and which is the bug).

To fix for row 3, we can convert the large number back to the default value.

| id | value1 | value2 | value1' | value2' | value1'' | value2'' | m=min(value1'', value2'') | m == 9999 ? 0 : m | exp_min
| - | - | - | - | - | - | - | - | - | -
| 1 | 100 | 200 | 100 | 200 | 100 | 200 | 100 | 100 | 100
| 2 | 100 | null | 100 | 0 | 100 | 9999 | 100 | 100 | 100
| 3 | null | null | 0 | 0 | 9999 | 9999 | 9999 | 0 | 0
| 4 | 100 | 0 | 100 | 0 | 100 | 9999 | 100 | 100 | 0

This also affects the TPP implementation.  (I have a failing test case.)

Reply:
To summarise, if we're finding the minimum or maximum of a set of values, then any values that match the default value will be ignored, because we also use the default value instead of NULL.  This is most likely to affect finding the minimum values, because the default value is either the empty string (which is less than every other string), or zero (and most values, eg blood pressure, counts of things) are positive.

@evansd any bright ideas?

Reply:
[According to @evansd on Slack](https://ebmdatalab.slack.com/archives/C63UXGB8E/p1608211898093500), this is fine.  I have updated the comment.